why-cant-we.md

```diff
@@ -0,0 +1,240 @@
+```
+
+Or in a type alias:
+
+```python
+type Movie = {"name": str, "year": int}
+```
+
+As a bonus, this also allows using keys that are not valid identifiers,
+like the existing call-based syntax for creating a TypedDict.
+
+The `NotRequired` and `ReadOnly` type qualifiers could also be used,
+e.g. `{"name": str, "year": NotRequired[int]}`.
+
+Note that some of the below problems can be mitigated by wrapping
+the dictionary, e.g. `TypedDict[{"name": str}]`. The in-preparation
+PEP 764 will propose this syntax.
+
+### Problems
+
+_`|` operator_: The `|` operator is already defined for dictionaries,
+and its behavior doesn't match the expected behavior for unions in
+the type system:
+
+```pycon
+>>> {"a": int} | {"b": str}
+{'a': <class 'int'>, 'b': <class 'str'>}
+>>> {"a": int} | {"a": str}
+{'a': <class 'str'>}
+```
+
+This would make it difficult to introspect annotations using unions
+of inline TypedDicts at runtime.
+
+Creating a union of an inline TypedDict and a type object currently
+throws an error:
+
+```pycon
+>>> {"a": int} | int
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+TypeError: unsupported operand type(s) for |: 'dict' and 'type'
+```
+
+However, this is fixable by changing the implementation of `__or__`
+on type objects.
+
+_Inheritance_: How would the inline syntax support inheriting from another
+TypedDict? The most obvious syntax that comes to mind is
+`{"name": str, **BaseMovie}`, where `BaseMovie` is another TypedDict.
+We'd want this to work both with TypedDicts created through `typing.TypedDict`
+and with type aliases like `type Movie = {...}`. The `**` operator works by
+calling `.keys()` and then calling `.__getitem__()` for each key. But the
+latter operation already has a meaning; it specializes a generic TypedDict
+or type alias.
+
+_TypedDict keyword arguments_: The class syntax for `TypedDict` currently
+supports one keyword argument, `total=False`, and the draft
+[PEP 728](https://peps.python.org/pep-0728/) proposes adding more. It is not
+obvious how these keyword arguments would be represented in the inline syntax.
+
+_Generics_: If a generic inline TypedDict is defined using the old `TypeAlias`
+syntax, like `MyDict: TypeAlias = {"a": T}`, it is impossible to specialize
+the generic at runtime; `MyDict[int]` would throw KeyError. However, the native
+syntax for type aliases in Python 3.12 fixes this; you could write
+`type MyDict[T] = {"a": T}` and `MyDict[int]` would work as expected, because
+subscripting gets handled by the type alias object, not the dict. Still, this could be
+a problem for users who want to use the inline syntax and maintain compatibility
+with Python 3.11 and earlier.
+
+## Tuples as tuple types
+
+### Idea
+
+Allow writing `(int, str)` as a type instead of `tuple[int, str]`.
+A single-element tuple type would be written as `(int,)`. Variadic
+tuple types (`tuple[int, ...]`) could potentially be written as
+`(int, ...)`.
+
+### Problems
```

Comment:
This also runs into problems with the `|` operator, similar to using `{}` for `TypedDict`s. We could add `tuple.__or__`, but that would make no sense at runtime outside of type annotations; it would be really baffling if we had this behaviour:

```pycon
>>> (int, str) | (int, bytes)
(int, str) | (int, bytes)
```

or worse:

```pycon
>>> (int, str) | (int, bytes)
tuple[int, str] | tuple[int, bytes]
```

Reply:
Thanks, good point! I'll add this and also add the "bare literals".